### PR TITLE
5867 add tjbf layer

### LINKED ISSUE #5867: Fully differentiable and trainable Joint Bilateral Filter (TJBF) layer
**Is your feature request related to a problem? Please describe.**
The [joint bilateral filter (JBF)](https://doi.org/10.1145/1015706.1015777) is similar to the bilateral filter but calculates its intensity range kernel on a second input image -- a so-called guidance image. This allows to integrate more information into the filter process and has slightly different applications than the usual bilateral filter (because you need a second guidance image for the filtering). MONAI currently does not contain a  regular or differentiable JBF.

**Describe the solution you'd like**
I implemented a fully differentiable and trainable version of the JBF based on an analytical filter gradient (C++/CUDA). You can find the code [here](https://github.com/faebstn96/trainable-joint-bilateral-filter-source). We additionally published a [paper (Scientific Reports)](https://doi.org/10.1038/s41598-022-22530-4) presenting and using that trainable joint bilateral filter for CT denoising.
The structure of the code is very similar to #1840 but contains the JBF algorithm with gradients with respect to its input, its guidance image, and its filter parameters. My GitHub also contains a gradcheck script to compare the analytical gradient to a numerical gradient based on `torch.autograd.gradcheck`.

I am happy to create a pull request similar to #5757 and include the trainable JBF as an additional filter layer. Please let me know what you think and add any questions or comments!

### PR DESCRIPTION
**Pull request by faebstn96**
Fixes #5867.

### Description

I integrated the trainable joint bilateral filter layer (TJBF) in the MONAI repository as a new PyTorch filter layer. The TJBF contains an analytical gradient derivation toward its filter parameters, its guidance image, and its noisy input image which enables gradient-based optimization within the PyTorch graph. See [here](https://doi.org/10.1038/s41598-022-22530-4) for more details on the gradient derivation. Unit tests were added that check the filter output as well as the gradient computations.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Review by wyli — COMMENTED** [review]
thanks, do you think the component names should have `_3d` in them?  all the logic seems to be coupled with three spatial dimensions.

e.g. https://github.com/Project-MONAI/MONAI/blob/30b2b11faa624e26224288f526432ba6df7446e6/monai/csrc/filtering/trainable_joint_bilateral/trainable_joint_bilateral.cpp#L38-L44

https://github.com/Project-MONAI/MONAI/blob/30b2b11faa624e26224288f526432ba6df7446e6/monai/networks/layers/filtering.py#L117-L121

File: monai/csrc/filtering/trainable_joint_bilateral/jbf_layer_cpu_backward.cpp [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "trainable_joint_bilateral.h"
+#include "utils/tensor_description.h"
+
+struct Indexer {
+ public:
```
[the remark this answers is not in the document]
Comment (faebstn96) [reply monai/csrc/filtering/trainable_joint_bilateral/jbf_layer_cpu_backward.cpp]: I moved the Indexer into utils/tensor_indexing.h to unify its definition. This should be better now, right?

[the remark this answers is not in the document]
Comment (wyli) [reply monai/csrc/filtering/trainable_joint_bilateral/jbf_layer_cpu_backward.cpp]: yes, thank you!

**Comment by faebstn96** [comment]
Do you mean renaming them to, e.g., sigma_3d_x etc? If you think that makes the code more readable I can do that.

**Comment by wyli** [comment]
I think in general the question is, as a user looking at this set of components https://github.com/Project-MONAI/MONAI/blob/2bcb52c188a4d6dd7ecc951b325b03bea2367e3f/monai/networks/layers/filtering.py#L18-L20
the underlying concepts are very similar but the input argument styles are different... 

ideally the signatures should be unified, but maybe that's out of the scope of this PR. so I'm thinking alternatively if we could rename the new components as `BilateralFilter3D` and `JointBilateralFilter3D` (same for the c++ function names), and in the docstring clarify the implementation differences among the components in file monai/networks/layers/filtering.py

**Comment by faebstn96** [comment]
So, the inputs to the `TrainableBilateralFilter` and `TrainableJointBilateralFilter` can also be [B, C, X] or [B, C, X, Y] tensors as there is some unsqueezing done:
https://github.com/Project-MONAI/MONAI/blob/5c778ad08c49b37c3764604c6fa56fcbd6dcb47f/monai/networks/layers/filtering.py#L234-L238
However, you are right, the underlying C++/Cuda code operates on three dimensions.
Though, I just recognized we should maybe make the spatial sigma parameters a tuple input like `sigma_spatial=(sigma_x, sigma_y, sigma_z)` and then check during layer initialization if `len(sigma_spatial)` equals the spatial dimensions of the input image/guidance tensors, right? This would make the layer flexible for 1D, 2D, and 3D inputs.

**Comment by wyli** [comment]
agreed, if it's renamed to `spatial_sigma`, then it's consistent with the existing `BilateralFilter`

**Comment by faebstn96** [comment]
Great, I just refactored the spatial_sigma initializations accordingly.

**Review by wyli — APPROVED** [review]
thanks, I'll merge this once the tests are ok (I'll push a workaround for the premerge test issue https://github.com/Project-MONAI/MONAI/issues/5859)

**Pull request merged**